File: internal/protocol/contract_test.go
package protocol

import (
	"bytes"
	"encoding/json"
	"os"
	"reflect"
	"sort"
	"testing"
)

// The wire protocol is mirrored, not generated (see AGENTS.md): protocol.go is
// canonical and web/src/lib/protocol/index.ts + fixture.ts are hand-written
// twins. testdata/snapshot_golden.json is the single captured contract both
// languages verify against. This test pins the Go structs to the golden; the
// TS test (web/src/lib/protocol/contract.test.ts) pins the TS twin + fixture to
// the same file. Rename/remove/add a field on one side and one of the two fails
// loud in CI instead of silently shipping a snapshot the other side can't read.

// collectPaths walks decoded JSON and returns the set of object field paths,
// with array indices normalized away (e.g. "spaces.tabs.panes.status"), so the
// key inventory can be compared structurally regardless of element count.
func collectPaths(v any, prefix string, set map[string]bool) {
	switch t := v.(type) {
	case map[string]any:
		for k, child := range t {
			p := k
			if prefix != "" {
				p = prefix + "." + k
			}
			set[p] = true
			collectPaths(child, p, set)
		}
	case []any:
		for _, e := range t {
			collectPaths(e, prefix, set)
		}
	}
}

func pathsOf(t *testing.T, raw []byte) map[string]bool {
	t.Helper()
	var m any
	if err := json.Unmarshal(raw, &m); err != nil {
		t.Fatalf("unmarshal to map: %v", err)
	}
	set := map[string]bool{}
	collectPaths(m, "", set)
	return set
}

func TestSnapshotContract(t *testing.T) {
	golden, err := os.ReadFile("testdata/snapshot_golden.json")
	if err != nil {
		t.Fatalf("read golden: %v", err)
	}

	// 1. The golden must decode into Snapshot with no unknown keys — catches a
	// field the golden (contract) carries that the Go structs no longer know.
	dec := json.NewDecoder(bytes.NewReader(golden))
	dec.DisallowUnknownFields()
	var snap Snapshot
	if err := dec.Decode(&snap); err != nil {
		t.Fatalf("golden has a field the Go structs reject: %v", err)
	}

	// 2. Re-marshalling the decoded value must reproduce exactly the golden's
	// key inventory — catches a field the Go structs emit that the golden lacks.
	out, err := json.Marshal(snap)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	goSet := pathsOf(t, out)
	goldenSet := pathsOf(t, golden)
	if !reflect.DeepEqual(goSet, goldenSet) {
		t.Fatalf("Go snapshot shape drifted from golden\n  go:     %v\n  golden: %v", sortedKeys(goSet), sortedKeys(goldenSet))
	}
}

func sortedKeys(m map[string]bool) []string {
	ks := make([]string, 0, len(m))
	for k := range m {
		ks = append(ks, k)
	}
	sort.Strings(ks)
	return ks
}

File: web/src/lib/protocol/contract.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { FIXTURE_SNAPSHOT } from '$lib/transport/fixture';

// The wire protocol is mirrored, not generated (AGENTS.md): the Go structs in
// internal/protocol are canonical and this TS twin + the fixture are hand-written.
// internal/protocol/testdata/snapshot_golden.json is the single captured contract.
// The Go test (internal/protocol/contract_test.go) pins the Go structs to it;
// this test pins the TS twin + fixture to the same file. FIXTURE_SNAPSHOT is
// typed `Snapshot`, so svelte-check already ties it to the TS types — asserting
// fixture ≡ golden therefore also ties the TS types to the golden. Drift on
// either side fails loud instead of shipping an unreadable snapshot.

// Walk decoded JSON collecting object field paths, array indices normalized away
// (e.g. "spaces.tabs.panes.status"), so key inventories compare structurally.
function collectPaths(v: unknown, prefix: string, set: Set<string>): void {
  if (Array.isArray(v)) {
    for (const e of v) collectPaths(e, prefix, set);
  } else if (v !== null && typeof v === 'object') {
    for (const [k, child] of Object.entries(v)) {
      const p = prefix ? `${prefix}.${k}` : k;
      set.add(p);
      collectPaths(child, p, set);
    }
  }
}

function pathsOf(v: unknown): Set<string> {
  const set = new Set<string>();
  collectPaths(v, '', set);
  return set;
}

describe('protocol wire contract', () => {
  const goldenPath = resolve(process.cwd(), '../internal/protocol/testdata/snapshot_golden.json');
  const golden = JSON.parse(readFileSync(goldenPath, 'utf8'));

  it('fixture shape matches the golden snapshot (minus the wire-only `type`)', () => {
    const goldenPaths = pathsOf(golden);
    goldenPaths.delete('type'); // TS `Snapshot` omits it; the SessionEvent union adds it on the wire
    const fixturePaths = pathsOf(FIXTURE_SNAPSHOT);
    expect([...fixturePaths].sort()).toEqual([...goldenPaths].sort());
  });

  it('golden carries the wire-only `type` discriminator', () => {
    expect(golden.type).toBe('snapshot');
  });
});
